src/flag_gems/ops/index_select.py

```diff
@@ -62,3 +62,59 @@ def index_select(inp, dim, index):
         return out.permute(order)
     else:
         return out
+
+
+# kernel
+@libentry()
+@triton.autotune(
+    configs=runtime.get_triton_config("index_select_backward"), key=["M", "N"]
+)
+@triton.jit
+def index_select_backward_kernel(
+    grad, out, M, N, index, outN, BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr
+):
+    pid_x = tle.program_id(axis=0)
+    pid_y = tle.program_id(axis=1)
+    rows_offsets = pid_x * BLOCK_M + tl.arange(0, BLOCK_M)[:, None]
+    rows_mask = rows_offsets < M
+    cols_offsets = pid_y * BLOCK_N + tl.arange(0, BLOCK_N)
+    grad_mask = rows_mask and (cols_offsets < N)
+    out_mask = rows_mask and (cols_offsets < outN)
+    indices = tl.load(index + cols_offsets, mask=(cols_offsets < N), other=0)
+    grad_off = rows_offsets * N + cols_offsets[None, :]
+    out_off = rows_offsets * outN + indices[None, :]
+    selected = tl.load(grad + grad_off, mask=grad_mask, other=0.0)
+    tl.store(out + out_off, selected, mask=out_mask)
+
+
+# function
+def index_select_backward(grad, self_sizes, dim, index):
+    logging.debug("GEMS INDEX SELECT BACKWARD")
+    assert dim >= -len(self_sizes) and dim < len(self_sizes), "Invalid dim"
+    assert index.ndim <= 1, "Index should have dimension 1 or 0"
+    for i in index:
+        assert i >= 0 and i < self_sizes[dim], "Index out of range"
+    if index.ndim == 0:
+        index = index.unsqueeze(0)
+    index_shape = list(index.shape)
+    dim = dim % len(self_sizes)
+    grad_shape = list(grad.shape)
+    assert grad_shape[dim] == index_shape[0], "Index out of range"
+    grad = dim_compress(grad, dim)
```

Comment:
Function `dim_compress` move the specified dim to the inner-most dimension and make a contiguous tensor, which is designed to be used in reduction/scan operator when other dimensions are considered batch dimensions and the reduction operation is performed on each 1d sub tensors, which will be loaded and iterated over. Thus being contiguous is good for this case.

For the backward of `index_select`, it is actually the opposite. Dimension `dim` is used  as an indexing dimension. Then several `n-1`d subtensors are inserted into a zeros tensor. So, being contiguous on dimension `dim` is not what we want. We rather want it to be the outer-most dimension.

Reply:
Ok,we will change it.

Reply:
Hello, there is a precision issue when calling atomic_add, and I'm not sure how to resolve it.